src/lib.rs

```diff
@@ -0,0 +1,353 @@
+        (unsafe{*self.0}).p_filesz
+    }
+
+    /// Returns the size of the segment when in memory.
+    pub fn memsz(&self) -> Elf_Xword {
+        (unsafe{*self.0}).p_memsz
+    }
+
+    /// Returns the alignment of the segment.
+    pub fn align(&self) -> Elf_Xword {
+        (unsafe{*self.0}).p_align
+    }
+}
+
+impl Debug for ProgramHeader {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut to_write = String::from("ProgramHeader(");
+
+        let typ = self.typ();
+        let typ_str = match typ {
+            PT_NULL => "PT_NULL",
+            PT_LOAD => "PT_LOAD",
+            PT_DYNAMIC => "PT_DYNAMIC",
+            PT_INTERP => "PT_INTERP",
+            PT_NOTE => "PT_NOTE",
+            PT_SHLIB => "PT_SHLIB",
+            PT_PHDR => "PT_PHDR",
+            PT_TLS => "PT_TLS",
+            PT_LOOS => "PT_LOOS",
+            PT_HIOS => "PT_HIOS",
+            PT_LOPROC => "PT_LOPROC",
+            PT_HIPROC => "PT_HIPROC",
+            PT_GNU_EH_FRAME => "PT_GNU_EH_FRAME",
+            PT_GNU_RELRO => "PT_GNU_RELRO",
+            // Some operating systems define their own non-standard segment types.
+            _ => "other",
+        };
+        to_write.push_str(&format!("typ={} ({}), ", typ, typ_str));
+
+        let flags = self.flags();
+        let mut flag_strs = Vec::new();
+        if flags & PF_X != 0 {
+            flag_strs.push("PF_X");
+        }
+        if flags & PF_W != 0 {
+            flag_strs.push("PF_W");
+        }
+        if flags & PF_R != 0 {
+            flag_strs.push("PF_R");
+        }
+        if flags & PF_MASKPROC != 0 {
+            flag_strs.push("PF_MASKPROC");
+        }
+        to_write.push_str(&format!("flags=<{}>, ", flag_strs.join("|")));
+
+        to_write.push_str(&format!("offset=<0x{:x}>, ", self.offset()));
+        to_write.push_str(&format!("vaddr=<0x{:x}>, ", self.vaddr()));
+        to_write.push_str(&format!("paddr=<0x{:x}>, ", self.paddr()));
+        to_write.push_str(&format!("align=<0x{:x}>)", self.align()));
+
+        write!(f, "{}", to_write)
+    }
+}
+
+
+/// An iterator over the program headers of an `Object`.
+///
+/// Each program header describes an ELF segment loaded in the virtual adress space.
+pub struct ProgramHeaderIterator {
+    ptr: *const p_ffi::Elf_Phdr,    // Pointer to the next raw `Elf_Phdr`.
+    num: p_ffi::Elf_Half,           // How many left.
+}
+
+impl Iterator for ProgramHeaderIterator {
+    type Item = ProgramHeader;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.num {
+            0 => None,
+            _ => {
+                let ret = Some(ProgramHeader(self.ptr));
+                self.ptr = unsafe {self.ptr.offset(1)};
+                self.num -= 1;
+                ret
+            }
+        }
+    }
+}
+
+/// Returns a `Vec` of objects loaded into the current address space.
+/// The reason this interface returns a `Vec` and not some kind of iterator, is that the C function
```

Comment:
Something like "This returns a `Vec` because we have to read all the values from `dl_iterate_phdr` calls in one go (i.e. we have to ensure that calls to `dl_iterate_phdr` are not interspersed)."?

Reply:
The first sentence is fine, but the second doesn't make sense. C calls our callback once per header, and we have no way to interrupt this to offer the user an intermediate result. Makes sense?

Reply:
Because it's a callback?

Reply:
Right.

Reply:
OK, then I think I would simply delete the "the reason" sentence. It's not very important detail and (if I'm being brutal) it's a bit waffly at the moment. It's probably best that we only tell the user things they really need to know, and I don't think an API is a great place for this.